test/reduce-e2e/reduce_test.go

```diff
@@ -46,31 +49,44 @@ func (r *ReduceSuite) TestSimpleKeyedReducePipeline() {
 	// port forward source vertex(to publish messages)
 	defer w.VertexPodPortForward("in", 8443, dfv1.VertexHTTPSPort).
 		TerminateAllPodPortForwards()
-
-	// publish messages to source vertex, with event time starting from 60000
-	startTime := 60000
-	for i := 0; i < 300; i++ {
-		eventTime := strconv.Itoa(startTime + i*1000)
-
-		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("1")).WithHeader("X-Numaflow-Event-Time", eventTime).
-			Expect().
-			Status(204)
-		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("2")).WithHeader("X-Numaflow-Event-Time", eventTime).
-			Expect().
-			Status(204)
-		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("3")).WithHeader("X-Numaflow-Event-Time", eventTime).
-			Expect().
-			Status(204)
-	}
+	done := make(chan struct{})
+
+	go func() {
+		// publish messages to source vertex, with event time starting from 60000
+		startTime := 60000
+		for i := 0; true; i++ {
+			select {
+			case <-ctx.Done():
+				return
+			case <-done:
+				return
+			default:
+				eventTime := strconv.Itoa(startTime + i*1000)
+
+				HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("1")).WithHeader("X-Numaflow-Event-Time", eventTime).
+					Expect().
+					Status(204)
+				HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("2")).WithHeader("X-Numaflow-Event-Time", eventTime).
+					Expect().
+					Status(204)
+				HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("3")).WithHeader("X-Numaflow-Event-Time", eventTime).
+					Expect().
+					Status(204)
+			}
+		}
+	}()
 
 	// since the key can be even or odd and the window duration is 10s
 	// the sum should be 20(for even) and 40(for odd)
 	w.Expect().VertexPodLogContains("sink", "Payload -  40  Key -  odd  Start -  60000  End -  70000")
```

Comment:
I am a bit new to go routine. Could you help me understand why the Expect can pass with the above go routine added?

My understanding is that the VertexPodLogContains has a one minute timeout(the default value) and within that one minute, it keeps searching for logs and finding matches. Meanwhile the go routine above keeps sending new data to the pipeline until the `w.Expect`s pass and send a signal back to `done` channel, which stops the go routine.

If my understanding is correct, will we ever reach the `case <-ctx.Done()`? (since Expect times out in one minute, which is less than the context timeout value 5 minutes)

Reply:
I didn't know the Expect times out after 1 min :). 
That's the thing right, the test owner doesn't know about the Expect timing out, so it's better to have their own context timeout to avoid go routine leaks or tests getting stuck.

Reply:
> Interestingly I see the expected output in the logs 😄

Right. The failure seems coming from the test clean up. `reduce_test.go:256: Timeout after 1m0s waiting for pipeline pods terminating.` Maybe 1m is too short for terminating all pods. Let's see if such case happens more frequently and if so, adjust the timeout value.

Reply:
Got it. I think it makes sense to allow e2e test implementer to explicitly specify test timeout values as oppose to using a global one in Makefile. For some test cases, we can fail faster.